testcases/InstallUpstreamKernel.py

```diff
@@ -0,0 +1,112 @@
+from common.OpTestSystem import OpSystemState
+
+log = OpTestLogger.optest_logger_glob.get_logger(__name__)
+
+
+class InstallUpstreamKernel(unittest.TestCase):
+    def setUp(self):
+        self.conf = OpTestConfiguration.conf
+        self.cv_HOST = self.conf.host()
+        self.cv_SYSTEM = self.conf.system()
+        self.host_cmd_timeout = self.conf.args.host_cmd_timeout
+        self.repo = self.conf.args.git_repo
+        self.branch = self.conf.args.git_branch
+        self.hostip = self.conf.args.host_ip
+        self.user = self.conf.args.host_user
+        self.password = self.conf.args.host_password
+        self.home = self.conf.args.git_home
+        self.config = self.conf.args.git_repoconfig
+        self.config_path = self.conf.args.git_repoconfigpath
+        self.disk = self.conf.args.host_scratch_disk
+        self.patch = self.conf.args.git_patch
+        if self.config_path:
+            self.config = "olddefconfig"
+        if not self.repo:
+            self.fail("Provide git repo of kernel to install")
+        if not (self.hostip and self.user and self.password):
+            self.fail("Provide host ip user details refer, --host-{ip,user,password}")
+
+    def runTest(self):
+        if self.disk:
+            log.debug("Set given disk as default bootable disk")
+            self.cv_SYSTEM.goto_state(OpSystemState.PETITBOOT_SHELL)
+            self.cv_SYSTEM.sys_set_bootdev_no_override()
+            time.sleep(60)
+            cmd = 'blkid %s*' % self.disk
+            output = self.cv_SYSTEM.console.run_command(cmd)
+            uuid = output[0].split(':')[1].split('=')[1].replace("\"", "")
+            cmd = 'nvram --update-config "auto-boot?=true"'
+            output = self.cv_SYSTEM.console.run_command(cmd)
+            cmd = 'nvram --update-config petitboot,bootdevs=uuid:%s' % uuid
+            output = self.cv_SYSTEM.console.run_command(cmd)
+            cmd = 'nvram --print-config'
+            output = self.cv_SYSTEM.console.run_command(cmd)
+        self.cv_SYSTEM.goto_state(OpSystemState.OS)
+
+        try:
+            con = self.cv_SYSTEM.cv_HOST.get_ssh_connection()
+        except Exception:
+            con = self.cv_SYSTEM.console
+
+        try:
+            onlinecpus = int(con.run_command("lscpu --online -e|wc -l")[-1])
+        except Exception:
+            onlinecpus = 20
+        log.debug("Downloading linux kernel")
+        # Compile and install given kernel
+        con.run_command("[ -d %s ] || mkdir -p %s" % (self.home, self.home))
+        con.run_command("if [ -d %s/linux ];then rm -rf %s/linux;fi" % (self.home, self.home))
+        con.run_command("cd %s && git clone --depth 1  %s -b %s linux" % (self.home, self.repo, self.branch), timeout=self.host_cmd_timeout)
+        con.run_command("cd %s/linux" % self.home)
+        if self.patch:
+            log.debug("Applying given patch")
+            # TODO: make copy_test_file_to_host generic to take from any source directory
```

Comment:
I think this has fixed here b726823c63f92dd3e8a7b66728664d06b7abd0d1 by you. so you can use it from custom directory.